### PR TITLE
Fix exiting `bal test` when a program has listeners and running workers/async functions

### LINKED ISSUE #40934: [Bug]: Runtime test exceptions are not thrown after using the generated main method for testable jar
### Description

$subject.
The program execution exits with only the below error without any detailed exceptions.
`there are test failures`

Was able to reproduce,

- When configurable variable tokens were expired and there was a runtime exception.
- For object mocking with test double, this is reproducible for only multi-module packages. 

After using the runtime generated main method the error handling is not working as expected.

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.6.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #41213: [Bug]: `bal test` does not exit when a module has a listener declaration and a continuously running worker/asynchronous function
### Description

$title

### Steps to Reproduce

Create a Ballerina package and add the following code to a module (e.g. default module).
```bal
import ballerina/http;
import ballerina/lang.runtime;

listener http:Listener httpListener = new(8080);

function init() returns error? {
    _ = start infiniteLoop();
}

public isolated function infiniteLoop() {
    while true {
        runtime:sleep(1);
    }
}

public function intAdd(int a, int b) returns (int) {
    return a + b;
}

service / on httpListener {
    resource function get greeting() returns string { 
        return "Hello, World!"; 
    }
}

```
and add the following Ballerina test (or another Ballerina test) in the `tests` folder.
```bal
import ballerina/test;

@test:Config {}
function intAddTest() {
    test:assertEquals(intAdd(1, 3), 4);
}
```

Running `bal test` hangs giving the below output.
```
Running Tests

        baltestissue


                1 passing
                0 failing
                0 skipped

```

### Affected Version(s)

2201.7.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
Fixes #41213
Fixes #40934

PR also contains changes from https://github.com/ballerina-platform/ballerina-lang/pull/41005.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by Nadeeshan96** [comment]
When there is a test failure, still `bal test` hangs, need to fix that and add a test.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/InitMethodGen.java [hunk 1]
```diff
@@ -319,11 +319,10 @@ private BIRNode.BIRFunction generateExecuteFunction(BIRNode.BIRPackage pkg, bool
                 addCheckedInvocation(modExecFunc, pkg.packageID, MODULE_START_METHOD, retVarRef, boolRef);
 
         if (testExecuteFunc != null) {
-            lastBB = addCheckedInvocationWithArgs(modExecFunc, pkg.packageID, TEST_EXECUTE_METHOD, retVarRef, boolRef,
+            addCheckedInvocationWithArgs(modExecFunc, pkg.packageID, TEST_EXECUTE_METHOD, retVarRef, boolRef,
                     functionArgs, Collections.emptyList());
-        }
-
-        if (!serviceEPAvailable && !JvmPackageGen.isLangModule(pkg.packageID)) {
+            lastBB = addCheckedInvocationForExitCall(modExecFunc, retVarRef, boolRef, typeOwnerClass);
```
Comment (gabilang) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/InitMethodGen.java:324]: Do we need to call this always for the test? Won't module stop take care of it?

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ModuleGracefulStopTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.HashMap;
+
+/**
+ * Test class to test graceful stop during module test execution.
```
Comment (gabilang) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ModuleGracefulStopTest.java:33]:
```suggestion
 * Tests to verify graceful stop during module test execution.
```

**Comment by Nadeeshan96** [comment]
Merged changes from https://github.com/ballerina-platform/ballerina-lang/pull/41005.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/InitMethodGen.java (hunk 1 above)
Comment (Nadeeshan96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/InitMethodGen.java]: For `bal test`, I think after the tests are done, at the end of the `$moduleExecute` function, it is okay to gracefully exit like this always. After all the threads are poisoned, `LaunchUtils.stopListeners` and `$moduleStop` will be called.

**Review by gabilang — DISMISSED** [review]
LGTM

**Review by HindujaB — DISMISSED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Comment by Nadeeshan96** [comment]
After I synched the PR with the master ProfilerTest failed due to assertion in `"○ Instrumented function count: "` showing the now the instrumented functions is less than 1 from before. Hence loosened the assertion. Created https://github.com/ballerina-platform/ballerina-lang/issues/41250 to verify the test outputs of the Profiler.

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**